modules/storages/app/services/storages/storages/create_service.rb

```diff
@@ -0,0 +1,47 @@
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2012-2021 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+# The logic for creating storage was extracted from the controller and put into
+# a service: https://dev.to/joker666/ruby-on-rails-pattern-service-objects-b19
+# Purpose: create and persist a Storages::Storage record
+# Used by: Storages::Admin::StoragesController#create, could also be used by the
+# API in the future.
+# Reference: https://www.openproject.org/docs/development/concepts/contracted-services/
+# The comments here are also valid for the other *_service.rb files
+module Storages::Storages
+  class CreateService < ::BaseServices::Create
+    protected
+
+    # Override the "creator_id" parameter with the actual user.
+    # before_perform is called in the service before attributes are set.
+    def before_perform(params, _service_result)
+      params[:creator_id] = user.id
```

Comment:
A better place for this is in the `SetAttributesService` as done [for work packages](https://github.com/opf/openproject/blob/dev/app/services/work_packages/set_attributes_service.rb#L76-L78). In there, the value is only set if no value was passed in today. That way, passing in a different creator_id (or `creator`-object` for that matter) is not masked and the contracts should then raise an error if that value is anything but the current user.

Reply:
Done in 9bc93a9168.

I forced the creator to be the current user in `default_attributes` but reading your comment again, I think I made a mistake. The SetAttributesService should be dumb and ensure it creates a coherent object with all values filled, and then it's contract's duty to check if it is valid.

The problem with what I have done is that a user could set the creator to an invalid value, and it would silently be fixed to be the current_user without letting the user knowing he submitted wrong data.

@ulferts Correct? Should I change?

Reply:
changed in 0bb1bde152

Reply:
@cbliard the approach in the second commit is correct. Assign a value if none is assigned before (from the client) and let the contract catch the error. :+1: